base/reduce.jl

```diff
@@ -362,10 +367,105 @@ julia> reduce(*, [2; 3; 4]; init=-1)
 -24
 ```
 """
-reduce(op, itr; kw...) = mapreduce(identity, op, itr; kw...)
+function reduce(op, itr::T; kw...) where T
+    # Redispatch, adding traits
+    reduce(op, itr, eltype_or_default_eltype(itr), IteratorSize(T); kw...)
+end
+
+function reduce(op, itr, et, isize; kw...)
+    # Fallback: if nothing interesting is being done with the traits
+    # or the operation
+    return mapreduce(identity, op, itr; kw...)
+end
 
 reduce(op, a::Number) = a  # Do we want this?
 
+##### Operation specific reduce optimisations
+
+## vcat
+
+function reduce(::typeof(vcat), xs, T::Type{<:AbstractVector}, isize)
+    x_state = iterate(xs)
+    x_state === nothing && return T() # New empty instance
+    x1, state = x_state
+
+    ret = copy(x1)  # this is **Not** going to work for StaticArrays
+
+    hinted_size = 0
+    if !(isize isa SizeUnknown)
+        # Assume first element has representitive size, unless that would make this too large
+        SIZEHINT_CAP = 10^5
+        expected_size = length(xs)*length(x1)
+        if expected_size < SIZEHINT_CAP
+            hinted_size = expected_size
+            sizehint!(ret, hinted_size)
+        end
+    end
+
+    x_state = iterate(xs, state)
+    while(x_state !== nothing)
+        x, state =  x_state
+        append!(ret, x)
+        x_state = iterate(xs, state)
+    end
+
+    if length(ret) < hinted_size/2  # it is only allowable to be at most 2x to much memory
+        sizehint!(ret, length(ret))
+    end
+
+    return ret
+end
+
+## hcat
+
+function reduce(::typeof(hcat), xs, T::Type{<:AbstractVector}, isize::SizeUnknown)
+    x_state = iterate(xs)
+    x_state === nothing && return et() # New empty instance
+    x1, state = x_state
+
+    dim1_size = length(x1)
+    dim2_size = 1
+    ret_vec = copy(x1)  # this is **Not** going to work for StaticArrays
+
+    x_state = iterate(xs, state)
+
+    while(x_state !== nothing)
+        x, state =  x_state
+        append!(ret_vec, x)
+        dim2_size += 1
+        x_state = iterate(xs, state)
+    end
+
+    # Reshape will throw errors if anything was the wrong size
```

Comment:
Why would anything be the wrong size?

Reply:
`reduce(hcat, ([1,2,3], [1,2]))`

Reply:
OK, I think you meant when something was corrupt.

Reply:
I will amend the comment